### ISSUE TITLE
how to train yolov5 with only one annotation file

### ISSUE THREAD

**Issue by qixuxiang** [ultralytics/yolov5#6119]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and found no similar feature requests.

### Description

Hi, Thanks for your great code share!
I  have millions of images for train, and the data is mounted to the cluster. But if I generate so many annotation files, the cluster cannot be trained, because my cluster is limited by the number of files.
I generated one annotation txt file like follow format (per line-> one image).
```
img_path1 x1,y1,x2,y2,class_id x1,y1,x2,y2,class_id2

img_path2 x1,y1,x2,y2,class_id
```
I would be very grateful if you could provide any useful suggestions, Thanks a lot!

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by glenn-jocher** [comment]
@qixuxiang see Train Custom Data tutorial for dataset formatting directions. 

### 1.2 Create Labels

After using a tool like [Roboflow Annotate](https://roboflow.com/annotate?ref=ultralytics) to label your images, export your labels to **YOLO format**, with one `*.txt` file per image (if no objects in image, no `*.txt` file is required). The `*.txt` file specifications are:

- One row per object
- Each row is `class x_center y_center width height` format.
- Box coordinates must be in **normalized xywh** format (from 0 - 1). If your boxes are in pixels, divide `x_center` and `width` by image width, and `y_center` and `height` by image height.
- Class numbers are zero-indexed (start from 0).

<img width="800" alt="Image Labels" src="https://user-images.githubusercontent.com/26833433/91506361-c7965000-e886-11ea-8291-c72b98c25eec.jpg">

The label file corresponding to the above image contains 2 persons (class `0`) and a tie (class `27`):

<p align="center"><img width="428" src="https://user-images.githubusercontent.com/26833433/112467037-d2568c00-8d66-11eb-8796-55402ac0d62f.png"></p>


### 1.3 Organize Directories

Organize your train and val images and labels according to the example below. YOLOv5 assumes  `/coco128` is inside a `/datasets` directory **next to** the `/yolov5` directory. **YOLOv5 locates labels automatically for each image** by replacing the last instance of `/images/` in each image path with `/labels/`. For example:
```bash
../datasets/coco128/images/im0.jpg  # image
../datasets/coco128/labels/im0.txt  # label
```


## YOLOv5 Tutorials

* [Train Custom Data](https://docs.ultralytics.com/yolov5/tutorials/train_custom_data)&nbsp; 🚀 RECOMMENDED
* [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/)&nbsp; ☘️ RECOMMENDED
* [Weights & Biases Logging](https://github.com/ultralytics/yolov5/issues/1289)&nbsp; 🌟 NEW
* [Supervisely Ecosystem](https://docs.ultralytics.com/yolov5/tutorials/model_export8)&nbsp; 🌟 NEW
* [Multi-GPU Training](https://docs.ultralytics.com/yolov5/tutorials/multi_gpu_training)
* [PyTorch Hub](https://docs.ultralytics.com/yolov5/tutorials/pytorch_hub_model_loading)&nbsp; ⭐ NEW
* [TFLite, ONNX, CoreML, TensorRT Export](https://docs.ultralytics.com/yolov5/tutorials/model_export) 🚀
* [Test-Time Augmentation (TTA)](https://docs.ultralytics.com/yolov5/tutorials/test_time_augmentation)
* [Model Ensembling](https://docs.ultralytics.com/yolov5/tutorials/model_ensembling)
* [Model Pruning/Sparsity](https://docs.ultralytics.com/yolov5/tutorials/model_pruning_and_sparsity)
* [Hyperparameter Evolution](https://docs.ultralytics.com/yolov5/tutorials/hyperparameter_evolution)
* [Transfer Learning with Frozen Layers](https://docs.ultralytics.com/yolov5/tutorials/transfer_learning_with_frozen_layers)&nbsp; ⭐ NEW
* [TensorRT Deployment](https://github.com/wang-xinyu/tensorrtx)

Good luck 🍀 and let us know if you have any other questions!

**Comment by balasuriyaranganathan** [comment]
I am having annotations folder which are in format .xml for each image and also there is a separate folder which has the labels files which are in .txt folder 

```
<annotation>  
	  <folder>images</folder>  
	  <filename>004270306.jpg</filename>  
	  <path>C:\Users\Desktop\000306.jpg</path>  
	  <source>    
		    <database>Unknown</database>    
	  </source>  
	  <size>    
		    <width>640</width>    
		    <height>480</height>    
		    <depth>3</depth>    
	  </size>  
	  <segmented>0</segmented>  
	  <object>    
		    <name>flower</name>    
		    <pose>Unspecified</pose>    
		    <truncated>0</truncated>    
		    <difficult>0</difficult>    
		    <bndbox>      
			      <xmin>353</xmin>      
			      <ymin>155</ymin>      
			      <xmax>500</xmax>      
			      <ymax>311</ymax>      
		    </bndbox>    
	  </object>  
</annotation>
```

that is one of the file in annotation folder in .xml format (say image1.xml)

here is how the file for image1.txt in labels  folder
```

0 0.5760416666666667 0.625 0.01875 0.0125
1 0.7270833333333333 0.55625 0.125 0.09062500000000001
1 0.17395833333333333 0.47734375 0.19375 0.13593750000000002

```

this is in the format of .txt is there anything i am missing now i am no table to train by loadint the .txt file how do i run it

**Comment by glenn-jocher** [comment]
@balasuriyaranganathan to train with YOLOv5, you need to convert your annotations from the XML format to the YOLO format (.txt). The YOLO format requires one .txt file per image, containing the labels in the format: 

`class x_center y_center width height`

Here are the steps to convert your annotations to the YOLO format:

1. Iterate through each image annotation in the XML format.
2. Extract the object class, bounding box coordinates (xmin, ymin, xmax, ymax), and image dimensions (width, height).
3. Normalize the bounding box coordinates by dividing them by the image dimensions.
4. Write the label information in the YOLO format to a .txt file, with one line per object.

For example, based on your sample annotations:

Image1.xml (sample XML annotation):
```
<annotation>  
      <folder>images</folder>  
      <filename>004270306.jpg</filename>  
      <path>C:\Users\Desktop\000306.jpg</path>  
      ...  
      <object>    
            <name>flower</name>    
            ...    
            <bndbox>      
                  <xmin>353</xmin>      
                  <ymin>155</ymin>      
                  <xmax>500</xmax>      
                  <ymax>311</ymax>      
            </bndbox>    
      </object>  
</annotation>
```

Image1.txt (corresponding YOLO format annotation):
```
0 0.59140625 0.6234375 0.0625 0.1171875
```
In the above example, the flower class is labeled as `0`, and the normalized bounding box coordinates are `(0.59140625, 0.6234375)` for the center of the object, and `(0.0625, 0.1171875)` for the width and height.

Make sure you run this conversion for each image annotation, and save the corresponding .txt files in the labels folder. Once you have the annotations in the YOLO format, you can train your YOLOv5 model using the `train.py` script.

Let us know if you have any further questions!
